src/test/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseActionTest.java

```diff
@@ -0,0 +1,279 @@
+
+    dialogs = new Dialogs();
+
+    notifier = mock(Notifier.class);
+
+    event = mock(AnActionEvent.class);
+    doReturn(project).when(event).getProject();
+
+    moduleSelectionDialogHelper = spy(new DialogHelper<>(viewModel -> {
+      viewModel.selectedModule.set(viewModel.getModules()[0]);
+      return true;
+    }));
+    moduleSelectionDialogFactory = new DialogHelper.Factory<>(moduleSelectionDialogHelper, project);
+    dialogs.register(ModuleSelectionViewModel.class, moduleSelectionDialogFactory);
+
+    submissionDialogHelper = spy(new DialogHelper<>(viewModel -> {
+      viewModel.selectedGroup.set(viewModel.getAvailableGroups().get(0));
+      return true;
+    }));
+    submissionDialogFactory = new DialogHelper.Factory<>(submissionDialogHelper, project);
+    dialogs.register(SubmissionViewModel.class, submissionDialogFactory);
+
+    action = new SubmitExerciseAction(mainVmProvider, fileFinder, moduleSource, dialogs, notifier);
+  }
+
+  @Test
+  public void testSubmitExerciseAction() throws IOException {
+    action.actionPerformed(event);
+
+    ArgumentCaptor<Submission> submissionArg = ArgumentCaptor.forClass(Submission.class);
+    verify(exerciseDataSource).submit(submissionArg.capture());
+
+    Submission submission = submissionArg.getValue();
+    assertEquals(group, submission.getGroup());
+    assertEquals(exercise, submission.getExercise());
+
+    Map<String, Path> files = new HashMap<>();
+    files.put(fileKey, filePath);
+    assertThat(submission.getFiles(), is(files));
+
+    verifyNoInteractions(notifier);
+  }
+
+  @Test
+  public void testNotifiesOfMissingFile() throws FileDoesNotExistException, IOException {
+    doReturn(null).when(fileFinder).tryFindFile(modulePath, fileName);
+
+    action.actionPerformed(event);
+
+    verifyNoInteractions(submissionDialogHelper);
+    verify(exerciseDataSource, never()).submit(any());
+
+    ArgumentCaptor<MissingFileNotification> notificationArg =
+        ArgumentCaptor.forClass(MissingFileNotification.class);
+
+    verify(notifier).notify(notificationArg.capture(), eq(project));
+
+    MissingFileNotification notification = notificationArg.getValue();
+    assertEquals(fileName, notification.getFilename());
+    assertEquals(modulePath, notification.getPath());
+
+    verifyNoMoreInteractions(notifier);
+  }
+
+  @Test
+  public void testNotifiesExerciseNotSubmittable() throws IOException {
+    doReturn(new SubmissionInfo(1, new SubmittableFile[0]))
+        .when(exerciseDataSource)
+        .getSubmissionInfo(exercise);
+
+    action.actionPerformed(event);
+
+    verifyNoInteractions(moduleSelectionDialogHelper);
+    verifyNoInteractions(submissionDialogHelper);
+    verify(exerciseDataSource, never()).submit(any());
+
+    verify(notifier).notify(any(NotSubmittableNotification.class), eq(project));
+
+    verifyNoMoreInteractions(notifier);
+  }
+
+  @Test
+  public void testNotifiesOfNetworkError1() throws IOException {
```

Comment:
can we please try avoid using the numbers in the method names and writing out the difference between the test scenarios with words :wink:  (I know it's quite challenging sometimes)